### PR TITLE
Update GlotPress localization generation to include new Swift Packages

### PR DESCRIPTION
**Pull request by mokagio**
Fixes an issue discovered by @crazytonyli where various entries had [disappeared from the `.strings` source of truth](https://github.com/wordpress-mobile/WordPress-iOS/commit/77eb1ad26ae6a0797671727ef950b520279847a9#diff-236a1390429e10f018cd3b7250bac8ce595acf8003dc42b43a01b034c485ca6e) during the 25.2 code freeze. This was due to moving the corresponding libraries from being fetched via CocoaPods to being local Swift Packages.

This PR address the issue by updating the localization automation configuration. 

Internal ref pdnsEh-1OV-p2

## Testing

To test these changes, I created a [test branch](https://github.com/wordpress-mobile/WordPress-iOS/compare/25.2.0.0...mokagio/test-update-localizations-on-25.2.0.0?expand=1) that cherry picks the commits from here on top of the 25.2.0.0 tag, then run the automation. 693d6f994880067e91f90fdb82d5c3e104c1b2b8 is the resulting commit where you can see the various strings that were removed in 77eb1ad26ae6a0797671727ef950b520279847a9 being reintroduced.

_I'll admit that I didn't check each individual edit. But here are a few random samples I took:_

image

image

image

_And here is a check for a string coming from the [WordPressKit repo](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/3832252c43c3710403ced45e7e2bc77f74add8e4/Sources/CoreAPI/WordPressOrgXMLRPCApi.swift#L213):_

image

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/294">25.3</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/lanes/localization.rb
Comment (mokagio) [inline fastlane/lanes/localization.rb]: Supporting a custom path makes is much faster to test the changes locally.

_But beware! If the Gutenberg path you give is one where you run `bundle install`, the automation will also generate `.strings` files from the spec in `release-toolkit`_

File: fastlane/lanes/build.rb
Comment (mokagio) [inline fastlane/lanes/build.rb]: I tried to use `build_app` with various `skip_*` flags, but couldn't get out of it running `xcodebuild build`.

I looked into the `xcodebuild` action, but it's a wrapper for canned commands that wouldn't have allowed to simply call this.

I made the derived data path a parameter with default because I find passing a value makes the localization code a bit clearer, see below.

File: fastlane/lanes/localization.rb
Comment (mokagio) [inline fastlane/lanes/localization.rb]: > I made the derived data path a parameter with default because I find passing a value makes the localization code a bit clearer, see below.

This is what I was referring to. I find using a dedicated value, `derived_data_path` between here and the paths list below makes the relationship between the two clear.

File: fastlane/lanes/localization.rb
Comment (mokagio) [inline fastlane/lanes/localization.rb:138]: If/when we had that option, then I'd have something like 

```ruby
UI.user_error!("CI detected but no derived_data_path provided.") if is_ci && options[:derived_data_path].nil?
```

Then at the point where the CI script invokes this lane (or its caller),

```bash
DERIVED_DATA_PATH=./DerivedData
install_swift_dependencies "$DERIVED_DATA_PATH"

bundle exec fastlane ... "derived_data_path:$DERIVED_DATA_PATH"
```

**Review by crazytonyli — APPROVED** [review]
Looks good to me. It's a shame that there is no action to just resolve packages in fastlane, but I don't see any issue in calling xcodebuild directly.

File: .buildkite/commands/complete-code-freeze.sh
Comment (AliSoftware) [inline .buildkite/commands/complete-code-freeze.sh:20]: 🎗️ Side-note (not directly related to this PR): WPiOS might be a good candidate for us testing the WIP improvements we're working on for this helper in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/105 & https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/108

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb]: 🎗️ Maybe worth creating an action for this in `release-toolkit` at some point (in a future, separate PR, of course)

That being said:
 - ⚠️  If I'm not mistaken, there's a risk for this invocation to update the package versions and the `Package.resolved` file (i.e. akin to `pod update` instead of `pod install`). You might want to also pass the `-onlyUsePackageVersionsFromResolvedFile` flag to this invocation to prevent this from happening.
 - ❓  I wonder if this lane and call is really needed in practice, given that the only lane that calls this is `generate_strings_file_for_glotpress`, and that lane is supposed to be run from CI, which already calls our `install_swiftpm_dependencies` helper from `ci-toolkit` before running the lane.
    - I guess it still makes sense to call it from the lane just as a way to ensure running a lane locally (e.g. for debugging purposes) would always result in the same outcome as running it on CI
    - OTOH, if the goal is only to _resolve_ SPM dependencies (as in, locally install the versions _pinned_ in the `Package.resolved`, as opposed to potentially _updating_ some of the Swift packages to a more recent version), I'd expect that to have already been done from opening the project in Xcode before testing the lane locally?

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [reply fastlane/lanes/localization.rb:138]: > `# FIXME: We yet don't have an explicit way to set the derived data in CI, so we'll run it in CI, too`

🤔  This comment is a bit confusing to understand to me. "We don't have a way to set derived data in CI, so we're setting derived data in CI"? It's like saying "we can't do X, so we are doing X" 😅 

Can you expand on what you really meant by that comment here?

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb]: 📝  Note for later: this would be a great occasion to migrate this lane from `|options|` to keyword parameters (like you did for the `lane :resolve_packages` one).
(That being said I can totally understand that, given the urgency to fix the issue in time for `25.2.0.0` without risking to break stuff, this might be better as a follow-up PR instead though)

Comment (AliSoftware) [reply fastlane/lanes/localization.rb]: 💄 In the meantime, may I suggest:
```suggestion
      derived_data_path = options.fetch(:derived_data_path, DERIVED_DATA_PATH)
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:142]: 💄  var name typo:
```suggestion
      custom_gutenberg_path = options[:gutenberg_path]
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:144]: 💄 
```suggestion
        # It's simpler to ask the caller to give an absolute path than to make it absolute ourselves
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:148]: 💄 
```suggestion
        UI.user_error!("Could not find Gutenberg project at given path #{custom_gutenberg_path}") unless Dir.exist?(custom_gutenberg_path)
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:146]: 💄 
```suggestion
        UI.user_error!("Please provide gutenberg_path as an absolute path. Current value: #{custom_gutenberg_path}") unless File.absolute_path?(custom_gutenberg_path)
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:150]: 💄 
```suggestion
        gutenberg_path = custom_gutenberg_path
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb]: 💄 
```suggestion
        UI.message("Using Gutenberg from #{gutenberg_path} instead of cloning it...")
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:155]: ❤️   Thanks for leaving such explanations as breadcrumb comments, as always 👌

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:162]: ❓ Just to confirm: do we want to support `ref_node[:branch]` as well or not?

I'm guessing no, because I wouldn't recommend pointing to a branch to begin with—as those can disappear at any point once the corresponding PR is merged and GitHub will delete it—but still asking just to be sure we're in agreement with the needs.

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:131]: ❓  Is this comment really accurate?

From reading the rest of the code, it seems that we actually _don't_ jump in the tempdir immediately, but instead only do the `Dir.chdir` later in the code (line 173) when we're ready to clone the repo. So this comment seems misleading?

Besides, I'd suggest to also move the `Dir.mktmpdir` on line 173 just before the repo cloning, otherwise we will create a useless tempdir needlessly if a custom `gutenberg_path` option is provided and we don't need any cloning to begin with.

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [inline fastlane/lanes/localization.rb:178]: 💄 
```suggestion
            if ref_node[:tag]
```

**Review by AliSoftware — APPROVED** [review]
Mostly [💄]  comment or implementation details nitpicks, [❓] a couple of non-blocking questions, and a few [🎗️] notes for later.

Nothing really blocking or that couldn't be addressed in a follow-up PR, so approving to unblock given the urgency of the request and upcoming release.

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb]: ⚠️ Actually, I'd also recommend against using a string for passing the command, as this will make `sh` run the command through shell interpretation and interpolation, and thus make it subject to potential injection (accidental or not) and prone to issues with paths containing spaces.

It's safer to pass the command argument as separate arguments to `sh` instead to avoid those risks and pitfalls altogether:

```suggestion
  lane :resolve_packages do |derived_data_path: DERIVED_DATA_PATH|
    sh(
      'xcodebuild',
      '-resolvePackageDependencies', '-onlyUsePackageVersionsFromResolvedFile',
      '-workspace',  File.join(PROJECT_ROOT_FOLDER, 'WordPress.xcworkspace'),
      '-scheme', 'WordPress',
      '-derivedDataPath', derived_data_path
    )
  end
```

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [reply fastlane/lanes/localization.rb:138]: Oooooh `<epiphany>` I just realized what you meant after all.

The issue is that the `install_swiftpm_dependencies` ci-toolkit utility doesn't (yet) expose a way for us to provide a custom DerivedData where to install those dependencies to, so that means that when that was run from the `.sh` before calling fastlane the dependencies were installed in Xcode's default DerivedData while in the rest of this code invoked from Fastlane we might want to use a different DerivedData location `</epiphany>`

Ok that now makes more sense to me why you've re-implemented the logic again via `resolve_packages` lane below now, got it.

---

The good news a that SwiftPM always first create a bare mirror of the SPM repos (`git clone --bare …`) in `~/Library/Caches/org.swift.swiftpm/repositories` [^1] **regardless of the location specified by `xcodebuild` for DerivedData**; and _only then_ it does a proper checkout[^2] from that bare repo / local mirror repository into the specified DerivedData folder.

Which means that, as long as you use `-onlyUsePackageVersionsFromResolvedFile`, the checkout will be super fast (almost instant), **even if the DerivedData path was not the same as when we called `install_swiftpm_dependencies`** to restore the SPM cache, because it will create a checkout[^2] of the files in the DerivedData folder… from the **local** bare repository[^1] in `~/Library/Caches`, as opposed to redownloading them from the net.

So TL;DR the SPM _cache_ and s independent of the DerivedData path (only the _checkout_ to transform the bare repo from the cache into a checkout into DerivedData is)

[^1]: A bare git repo is basically only the content of the `.git` folder you're used to, with all the repo history and git objects/refs/…, but without the human-browsable files they represent in the filesystem.

[^2]: In this context, the term "git checkout" is used in the strict sense of the git terminology, and basically consists of git creating human-browsable files on the filesystem representing the state of the objects in the `.git/objects` database.

File: fastlane/lanes/build.rb
Comment (mokagio) [reply fastlane/lanes/build.rb]: @AliSoftware 

> OTOH, if the goal is only to resolve SPM dependencies (as in, locally install the versions pinned in the Package.resolved, as opposed to potentially updating some of the Swift packages to a more recent version), I'd expect that to have already been done from opening the project in Xcode before testing the lane locally?

Yes that's the goal and thank you for improving the implementation by suggesting the `-onlyUsePackageVersionsFromResolvedFile`.

The reason we need this automation is because the SPMs need to be checked out in CI, in particular in the lane that finalizes the code freeze. That lane does not build the app so we need to force fetching the sources of the SPM dependencies in the CI instance. So, while it's true that a dev should always have the sources in their machine, CI might not.

I made this a lane rather than a function or a call directly in the localization lane because it seemed handy to have it available, if just for testing.

File: fastlane/lanes/localization.rb
Comment (mokagio) [reply fastlane/lanes/localization.rb:138]: Good point. Here's my understanding:

We don't yet have a way to ensure that `install_swift_dependencies` and `generate_strings_file_for_glotpress` use the same derived data folder from CI. That's because `install_swift_dependencies` doesn't accept a value to pass to `xcodebuild` via `-derivedDataPath` ( 🤓🏹 )

If we could, I presume we would be able to bypass calling `resolve_packages` if `is_ci` and pass the path to the derived data used by `install_swift_dependencies`, where we would fine the SPM sources.

Since that's not possible yet, we need to resolve the packages in CI within the lane, just to make sure.

---

Update: When I wrote this comment, @AliSoftware 's [latest comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/23506#discussion_r1725934262) wasn't in my page 😓 

The thing about `xcodebuild` using `~/Library/Caches` was a TIL for me. 

I'll update the comment removing the FIXME and linking to the discussion here.

File: fastlane/lanes/localization.rb
Comment (mokagio) [reply fastlane/lanes/localization.rb:131]: >❓ Is this comment really accurate?
>
> From reading the rest of the code, it seems that we actually don't jump in the tempdir immediately,

That's correct. I even went back and did a `git commit --fixup` to replace "jump" with "create" but I clearly messed up a conflict resolution when interactive rebasing it...

**Comment by mokagio** [comment]
Thanks @AliSoftware for the many suggesionts here. 

I implemented some but run out of time to tackle all. I'll enable auto-merge so this can land before the 25.3 code freeze. I'll keep following up in the upcoming days.

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb]: Got it. Makes more sense to me now 👍 

I wonder how we could ensure that the implementation of this lane doesn't risk being out of sync with the implementation of the same feature in `bin/install_swiftpm_dependencies` (e.g. I'm especially thinking about keeping the list of flags used when calling `xcodebuild` in sync) though, to avoid inconsistencies potentially popping up in the future around this implementation detail.

Maybe the long term solution (future PR) to ensure a single source of truth and avoid implementations from drifting away would be:
 - Make `install_swiftpm_dependencies` be able to take an extra parameter to specify the derived data path where to checkout the packages
 - Pass such parameter when calling `install_swiftpm_dependencies` in the `.buildkite/*.sh` scripts of this repo
 - Remove the `resolve_packages` lane, and replace it with a `UI.user_error!("Before running this translation lane locally, please ensure that you have asked Xcode to resolve SPM package dependencies first") if !is_ci && (code to check if package checkouts already exist in DerivedData path or not)`? 🤔 

[EDIT] Wrote this before seeing [your comment below](https://github.com/wordpress-mobile/WordPress-iOS/pull/23506#discussion_r1726823515) which basically suggests the same thing for a long term solution. Great minds… 🙂

File: fastlane/lanes/localization.rb
Comment (AliSoftware) [reply fastlane/lanes/localization.rb:138]: > The thing about xcodebuild using ~/Library/Caches was a TIL for me.

Yep, I learned this (+ a lot about git repositories vs checkouts) not so long ago while I was looking a bit more in details how SwiftPM caching logic worked while working on https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/105 (and because I was wondering if the DerivedData used on CI would affect the location we'd need to cache in our `install_swiftpm_dependencies` implementation)

I actually plan to write about these in a P2 soon.

File: fastlane/lanes/localization.rb
Comment (mokagio) [reply fastlane/lanes/localization.rb:162]: Confirming that I don't think we should support reading from a branch for the very same rationale you gave.

Semi-related internal discussion pbArwn-6DH-p2

**Pull request merged**
